### PR TITLE
Fix: incorrect default number of products in editor when inheriting query

### LINKED ISSUE #10228: Product Collection - incorrect default number of products in editor when inheriting query
### Description
By default when blocks inherit from query templates, there's a difference in number of displayed products:

### Steps to reproduce

1. Go to Editor and edit Product Catalog template
2. Insert Product Collection and Products (Beta) blocks
3. Make sure both of them have "Inherit query from template" options enabled
4. Save the template
5. Check number of products displayed in editor and frontend for both blocks:

x | Products (Beta) | Product Collection
-- | -- | --
Editor | 10 | 9
Frontend | 12 | 12

### Expected outcome
Template query is controlled externally, hence we should adjust to that. If it displays 12 products, both Products (Beta) and Product Collection block should display this amount in Editor as well, so:

- when "Inherit query from template" option is enabled items per page should be 12.
- otherwise, it should be 9 by default.

### PR DESCRIPTION
**Pull request by dinhtungdu**
Fixes #10228

In this PR, we're using two different approaches to fix the same problem:
- For `Products (Beta)` block, as we're extending the Query Loop block, we expose the `__woocommerceInherit` parameter to the REST query then modify query using the `rest_product_query` filter to inject the correct loop items per page value to the `posts_per_page` argument.
- For `Product Collection` block, because we have total control over the codebase, we're following what Gutenberg does with the Query Loop block by setting the correct loop items per page directly in the edit component.

#### Other Checks

- [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] This PR adds/removes an experimental interfaces and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] I tagged two reviewers because this PR makes queries to the database or I think it might have some security impact.

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests

#### User Facing Testing

1. Go to Editor and edit Product Catalog template
2. Insert Product Collection and Products (Beta) blocks
3. Make sure both of them have "Inherit query from template" options enabled
4. Save the template
5. Check number of products displayed in editor and frontend for both blocks:

x | Products (Beta) | Product Collection
-- | -- | --
Editor | 12 | 12
Frontend | 12 | 12

* [ ] Do not include in the Testing Notes 

### WooCommerce Visibility

* [x] WooCommerce Core
* [ ] Feature plugin
* [x] Experimental

Because this PR updates the `Products (Beta)`, I checked the box for `WooCommerce Core` too.

### Performance Impact

N/a

### Changelog

> Fix: incorrect default number of products in editor when inheriting query

### CODE REVIEW

File: assets/js/blocks/product-query/inspector-controls.tsx [hunk 1]
```diff
@@ -139,6 +139,7 @@ export const WooInheritToggleControl = (
 				setQueryAttribute( props, {
 					...props.defaultWooQueryParams,
 					inherit,
+					__woocommerceInherit: inherit,
```
Comment (imanish003) [inline assets/js/blocks/product-query/inspector-controls.tsx]: AFAIK, we should use `inherit` instead of `__woocommerceInherit`. The `__woocommerceInherit` is added only when `isCustomInheritGlobalQueryImplementationEnabled` is true, but we've decided not to implement the custom inherit query from template. Therefore, this flag will always be false, as seen [here](https://github.com/woocommerce/woocommerce-blocks/blob/adc72972bf76b0e2dee7ecf6a9ab25c2f7085639/assets/js/blocks/product-query/utils.tsx#L63). You can also find related discussion on Slack here: p1681812585287949-slack-C02UBB1EPEF.

In short, we shouldn’t use `__woocommerceInherit`. Additionally, I believe we can remove the `isCustomInheritGlobalQueryImplementationEnabled` flag and related code since we decided not to implement the "custom inherit query from template." I understand it might be confusing because there is still a partial implementation for the "custom inherit query from template," which is disabled by using the `isCustomInheritGlobalQueryImplementationEnabled` flag.

Does that make sense?

CC: @gigitux

**Review by imanish003 — CHANGES_REQUESTED** [review]
Hi @dinhtungdu,

Thanks for working on this. I was verifying the number of products as mentioned in the following table:

image

However, for the Product (Beta) block, the number of products is not correct in the Editor. It's showing 9 products instead of 12 in the editor. But, it displays the correct number of products on the frontend. I'm not sure if this is happening only for me. 🤷🏻‍♂️

File: assets/js/blocks/product-query/inspector-controls.tsx (hunk 1 above)
Comment (gigitux) [reply assets/js/blocks/product-query/inspector-controls.tsx]: > In short, we shouldn’t use __woocommerceInherit. Additionally, I believe we can remove the isCustomInheritGlobalQueryImplementationEnabled flag and related code since we decided not to implement the "custom inherit query from template." I understand it might be confusing because there is still a partial implementation for the "custom inherit query from template," which is disabled by using the isCustomInheritGlobalQueryImplementationEnabled flag.

Since the logic is added on the Products block (that will be deprecated) I agree with you, we can remove that logic!

Comment (dinhtungdu) [reply assets/js/blocks/product-query/inspector-controls.tsx]: Thanks @imanish003 for the review, I agree this is a bit confusing, so let's discuss more.

- In [0d8fc18](https://github.com/woocommerce/woocommerce-blocks/pull/10303/commits/0d8fc1833cfe8f60015be5570f04346badab7828), I fixed the issue you mentioned in the review. I didn't handle well the case adding the block into archive template in the template editor.
- I use the `__woocommerceInherit` attribute just for detecting if the current rest query is inherited from template. I used it in the `rest_product_query` filter [here](https://github.com/woocommerce/woocommerce-blocks/pull/10303/files#diff-efb947e345af5e501dfaa2980b055b3b31c308b31e8db23049b6b48fecee33d8R197) (as a custom [rest query argument](https://github.com/WordPress/gutenberg/blob/81b58d30a8685872fd5ac361bb75760d50318e1d/packages/block-library/src/post-template/edit.js#L190)). The reason for that is I can't find a better way, I'm open to any alternative that we can detect the 'inherit state' in the rest query. I admit this is a workaround than a solution, but given that the `Products (Beta)` block will be deprecated, this is acceptable IMO.
- About removing `isCustomInheritGlobalQueryImplementationEnabled`, let's do it in a separated PR for better history.

Comment (imanish003) [reply assets/js/blocks/product-query/inspector-controls.tsx]: > In [0d8fc18](https://github.com/woocommerce/woocommerce-blocks/pull/10303/commits/0d8fc1833cfe8f60015be5570f04346badab7828), I fixed the issue you mentioned in the review. I didn't handle well the case adding the block into archive template in the template editor.

Not sure why but it's still showing 9 products in Products beta block 🤔  Are you able to replicate this?

> I use the __woocommerceInherit attribute just for detecting if the current rest query is inherited from template. I used it in the rest_product_query filter [here](https://github.com/woocommerce/woocommerce-blocks/pull/10303/files#diff-efb947e345af5e501dfaa2980b055b3b31c308b31e8db23049b6b48fecee33d8R197) (as a custom [rest query argument](https://github.com/WordPress/gutenberg/blob/81b58d30a8685872fd5ac361bb75760d50318e1d/packages/block-library/src/post-template/edit.js#L190)). The reason for that is I can't find a better way, I'm open to any alternative that we can detect the 'inherit state' in the rest query. I admit this is a workaround than a solution, but given that the Products (Beta) block will be deprecated, this is acceptable IMO.

Got it. In that case, maybe we should consider naming this something other than __woocommerceInherit because its original intention was for use in the "custom inherit from query" implementation. Since this fix is not directly related to "custom inherit from query," using __woocommerceInherit for this fix might cause confusion. What do you think?

> About removing isCustomInheritGlobalQueryImplementationEnabled, let's do it in a separated PR for better history.

Definitely, Also this doesn't fall into scope of this PR.

Comment (dinhtungdu) [reply assets/js/blocks/product-query/inspector-controls.tsx]: > Since this fix is not directly related to "custom inherit from query," using __woocommerceInherit for this fix might cause confusion. What do you think?

@imanish003 Because we're removing that "custom inherit feature", and the Products (Beta) block will be deprecated eventually, I don't think that's a big issue. But by the way, any name will work, I will try to find another one to clear any potential confusion.

> Not sure why but it's still showing 9 products in Products beta block 🤔 Are you able to replicate this?

Let's pair on this.

Comment (imanish003) [reply assets/js/blocks/product-query/inspector-controls.tsx]: >  But by the way, any name will work, I will try to find another one to clear any potential confusion.

Thank you so much for your understanding. Since we're unsure who will be working on removing the "custom inherit query" feature, it's better to avoid any potential confusion if we can. 🙌🏻

**Comment by dinhtungdu** [comment]
@imanish003 I updated my PR to fix the issue with Products Beta block, can you please take another look at this?

**Review by imanish003 — APPROVED** [review]
Awesome work! Tested and confirmed that the Products beta block in the Editor now displays 12 products. 🚀

**Comment by dinhtungdu** [comment]
Something to clarify:

- We only introduce a partial solution for the `Products (Beta)` block. If the site changes the number of products per page after setting the product archive templates with the `Products (Beta)` block, that number won't reflect in the editor. Users must re-add the block or toggle off and on the `Inherit query from template` to update the correct per-page value. This is because we are extending the `Query Loop` block, and that block uses the `Settings > Reading > Blog pages show at most` value for the number of posts per page, and we can't filter/change it.
- `Product Collection` however doesn't have that issue because we have total control over the code base, we take a similar approach that `Query Loop` uses to have the number of products per page (in the editor) dynamically set according to the correct value.

cc @imanish003
